Frameworks/CoreText/CTFrame.mm

```diff
@@ -113,28 +113,41 @@ void CTFrameDraw(CTFrameRef frameRef, CGContextRef ctx) {
     _CTFrame* frame = static_cast<_CTFrame*>(frameRef);
     if (frame && ctx) {
         // Need to invert and translate coordinates so frame draws from top-left
-        CGContextSaveGState(ctx);
         CGRect boundingRect = CGPathGetBoundingBox(frame->_path.get());
         CGContextTranslateCTM(ctx, 0, boundingRect.size.height);
 
-        // Invert Text Matrix and CTM so glyphs are drawn in correct orientation and correct position
+        // Invert Text Matrix and CTM so glyphs are drawn in correct orientation and position
         CGAffineTransform textMatrix = CGContextGetTextMatrix(ctx);
-        CGContextSetTextMatrix(
-            ctx, CGAffineTransformMake(textMatrix.a, -textMatrix.b, textMatrix.c, -textMatrix.d, textMatrix.tx, textMatrix.ty));
+        CGContextSetTextMatrix(ctx, CGAffineTransformMake(textMatrix.a, -textMatrix.b, textMatrix.c, -textMatrix.d, 0, 0));
         CGContextScaleCTM(ctx, 1.0f, -1.0f);
 
         _CGContextPushBeginDraw(ctx);
-        auto popEnd = wil::ScopeExit([ctx]() { _CGContextPopEndDraw(ctx); });
+        auto popEnd = wil::ScopeExit([&]() {
+            // Restore CTM and Text Matrix to values before we modified them
+            CGContextSetTextMatrix(ctx, textMatrix);
+            CGContextScaleCTM(ctx, 1.0f, -1.0f);
+            CGContextTranslateCTM(ctx, 0, -boundingRect.size.height);
+            _CGContextPopEndDraw(ctx);
+        });
+
+        std::vector<GlyphRunData> runs;
 
         for (size_t i = 0; i < frame->_lineOrigins.size() && (frame->_lineOrigins[i].y < frame->_frameRect.size.height); ++i) {
             _CTLine* line = static_cast<_CTLine*>([frame->_lines objectAtIndex:i]);
-            CGContextSetTextPosition(ctx, frame->_lineOrigins[i].x, frame->_lineOrigins[i].y);
-            CTLineDraw(static_cast<CTLineRef>(line), ctx);
+            CGPoint relativePosition = frame->_lineOrigins[i];
+            for (size_t j = 0; j < [line->_runs count]; ++j) {
+                _CTRun* curRun = [line->_runs objectAtIndex:j];
+                if (j > 0) {
+                    // Adjusts x position relative to the last run drawn
+                    relativePosition.x += curRun->_relativeXOffset;
+                }
+                runs.emplace_back(GlyphRunData{ &curRun->_dwriteGlyphRun, relativePosition, (CFDictionaryRef)curRun->_attributes.get() });
+            }
         }
 
-        // Restore CTM and Text Matrix to values before we modified them
-        CGContextRestoreGState(ctx);
-        CGContextSetTextMatrix(ctx, textMatrix);
+        if (!runs.empty()) {
+            _CGContextDrawGlyphRuns(ctx, runs.data(), runs.size());
```

Comment:
There's no point in doing a push begin and pop end if `runs.empty()`. You could concentrate all the drawing logic here.